### PR TITLE
:seedling: Add verify-import-restrictions to enforce import restrictions

### PR DESCRIPTION
**Pull request by laozc**
**What this PR does / why we need it**:
This PR adds `verify-import-restrictions` to enforce import restrictions on API packages
to prevent regression mentioned in #9011.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by sbueringer** [comment]
/approve

Looks good to me thx. Let's get this PR in first, then we can do the same for other packages

/assign @chrischdi @killianmuldoon

**Comment by mdbooth** [comment]
FYI, this looks like it duplicates the functionality of the golangci module [depguard](https://github.com/OpenPeeDeeP/depguard). I believe this is a standard module, as I had to disable it in CAPO when updating to golangci-lint-1.54.

**Comment by sbueringer** [comment]
I think we had depguard enabled in the past (but didn't actually use it) but dropped it for some reason.

Now we're using the same as k/k

@killianmuldoon @chrischdi Do you remember why we removed depguard from the enabled linter list?

**Comment by chrischdi** [comment]
> I think we had depguard enabled in the past (but didn't actually use it) but dropped it for some reason.
> 
> Now we're using the same as k/k
> 
> @killianmuldoon @chrischdi Do you remember why we removed depguard from the enabled linter list?

TLDR: we did not configure it and beginning with a golangci lint version it would have required us to add some config. So we removed the linter instead because we obviously did not use it.

Source:

> Removed `depguard` which now requires a default config with all of the dependencies allowed or disallowed.

https://github.com/kubernetes-sigs/cluster-api/pull/8796
